Unit attack, move and wait.

hilalcomak opened the pull request.

paolovic commented on src/main/java/ch/uzh/ifi/hase/soprafs22/rest/mapper/DTOMapper.java:
```diff
@@ -71,4 +73,7 @@ int convertTeamToTeamNumber(Team team){
     @Mapping(source = "gameMode", target = "gameMode")
     @Mapping(source = "gameMap", target = "gameMap")
     public abstract GameGetDTO convertGameToGameGetDTO(Game game);
+    public Position convertPositionDTOToPosition(PositionDTO position) {
```
abstract should be enough, shouldn't it?

hilalcomak replied:
Indeed, done.

hilalcomak replied:
Can't do. Position has no setters (and it doesnt need them) so we just build a new one.

paolovic commented on src/main/java/ch/uzh/ifi/hase/soprafs22/game/Game.java:
```diff
@@ -76,4 +78,86 @@ public boolean hasEnded(){
+        ensureMember(token);
+        ensureNotEnded();
+        ensureTurn(token);
+        Unit attackingUnit = gameMap.getTile(attacker).getUnit();
```
i changed that, but i'm open to let me convince me otherwise.

hilalcomak replied:
It is still usable, but finding units gets more complicated. For example, when checking if unit at position (x,y) belongs to an enemy or teammate we used to:
1. Get tile at (x, y).
2. Check the team of the unit.

Now we have to:
1. Find all players in the game.
2. Find all units of each player.
3. Check if one is at position (x,y).
4. Check the team of that unit.

I've changed it to the current implementation, please fix this if you decide to keep units in their tiles.

paolovic replied:
Good thought. I'll think about this.

apple00juice commented on src/main/java/ch/uzh/ifi/hase/soprafs22/service/GameService.java:
```diff
@@ -26,10 +27,123 @@
 public class GameService {
     private final Logger log = LoggerFactory.getLogger(GameService.class);
     private final UserRepository userRepository;
+    private final Map<Long, Game> gameMap = new HashMap<>();
```
We are storing the game in the lobby at the moment, so this would never be filled
eg. lobbymanger > get lobby > get game

hilalcomak replied:
Why are we storing games inside lobbies?

apple00juice replied:
I would say because there exists a one-to-one connection between lobbies and a game. But I am open for a gameManger or something like that.

paolovic replied:
> Why are we storing games inside lobbies?

because we set all the relevant game settings in the lobby and start the game from the lobby.

paolovic replied:
furthermore: gameMap collides with the name for the actual map of the game (where the units are placed and the game takes place etc pp) which is also named gameMap

hilalcomak replied:
imo it makes more sense that the lobby creates a game with the settings:

1. Lobbies should be deleted once promoted into games.
2. It makes no sense to interact with a lobby once the game has started.
3. All the mutable settings in lobby should be immutable in the game.

Keeping them together opens lots of questions/checks to add in lobby.
- Is changing ones name in the lobby OK while the game is ongoing?
- Can I leave a lobby and not leave the game?
- What happens to the lobby if I set the game to ranked?

As for a gameManager, there is no need for it, gameService is fine.

You decide, I can add lobbyService to gameService and get it from there but then everything will go through lobby and we will end up with a lobby god class.

paolovic replied:
Good thought. I'll think about this.

apple00juice reviewed (approved):
Rest looks good 👍

hilalcomak commented on src/main/java/ch/uzh/ifi/hase/soprafs22/controller/GameController.java:
```diff
@@ -18,11 +26,35 @@ public class GameController {
+    public void unitWait(@RequestHeader("token") String token, @PathVariable Long id, @RequestBody WaitPostDTO waitPostDTO) {
+        Position unitPosition = DTOMapper.INSTANCE.convertPositionDTOToPosition(waitPostDTO.getUnitPosition());
+
+        gameService.unitWait(id, token, unitPosition);
```
I did this because wait is a reserved word https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#wait(). Happy to change them

hilalcomak commented on src/main/java/ch/uzh/ifi/hase/soprafs22/exceptions/GameNotFoundException.java:
```diff
@@ -0,0 +1,14 @@
+package ch.uzh.ifi.hase.soprafs22.exceptions;
+
+public class GameNotFoundException extends Exception {
```
Thanks!

hilalcomak commented on src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/PositionDTO.java:
```diff
@@ -0,0 +1,22 @@
+package ch.uzh.ifi.hase.soprafs22.rest.dto;
+
+public class PositionDTO {
```
Position is an inside object used in other DTOs. We can reuse it for all of them.

hilalcomak commented on src/main/java/ch/uzh/ifi/hase/soprafs22/controller/GameController.java:
```diff
@@ -18,11 +26,35 @@ public class GameController {
+
+    @PostMapping("/{apiVersion}/game/match/{id}/command/move")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void move(@RequestHeader("token") String token, @PathVariable Long id, @RequestBody MovePostDTO movePostDTO) {
```
Done!

paolovic commented:
thx

hilalcomak commented:
@paolovic I changed the GameService according to the comment (retrieving Game from Lobby .) Also, removed the GameMap from GameService. What do you think?
